src/nbl/video/CCUDADevice.cpp

```diff
@@ -1,131 +1,159 @@
-	{
-		if (!iit->acquired)
-			return CUDA_ERROR_UNKNOWN;
+	auto& cu = m_handler->getCUDAFunctionTable();
 
-		result = cuda::CCUDAHandler::cuda.pcuGraphicsResourceGetMappedMipmappedArray(&iit->asImage.mipmappedArray,iit->cudaHandle);
-		if (result != CUDA_SUCCESS)
-			return result;
-	}
-	return CUDA_SUCCESS;
-}
-CUresult CCUDAHandler::acquireAndGetArray(GraphicsAPIObjLink<video::IGPUImage>* linksBegin, GraphicsAPIObjLink<video::IGPUImage>* linksEnd, uint32_t* arrayIndices, uint32_t* mipLevels, CUstream stream)
-{
-	if (linksBegin+MaxAquireOps<linksEnd)
-		return CUDA_ERROR_OUT_OF_MEMORY;
-	alignas(_NBL_SIMD_ALIGNMENT) uint8_t stackScratch[MaxAquireOps*sizeof(void*)];
+	uint32_t metaData[16] = { 48 };
 
-	CUresult result = acquireResourcesFromGraphics(stackScratch,linksBegin,linksEnd,stream);
-	if (result != CUDA_SUCCESS)
-		return result;
+	CUmemAllocationProp prop = {
+		.type = CU_MEM_ALLOCATION_TYPE_PINNED,
+		.requestedHandleTypes = ALLOCATION_HANDLE_TYPE,
+		.location = { .type = params.location, .id = m_handle },
+		.win32HandleMetaData = metaData,
+	};
+	
+	params.granularSize = roundToGranularity(params.location, params.size);
 
-	auto ait = arrayIndices;
-	auto mit = mipLevels;
-	for (auto iit=linksBegin; iit!=linksEnd; iit++,ait++,mit++)
+	CUmemGenericAllocationHandle mem;
+	if(auto err = cu.pcuMemCreate(&mem, params.granularSize, &prop, 0); CUDA_SUCCESS != err)
+		return err;
+	
+	if (auto err = cu.pcuMemExportToShareableHandle(&params.osHandle, mem, prop.requestedHandleTypes, 0); CUDA_SUCCESS != err)
 	{
-		if (!iit->acquired)
-			return CUDA_ERROR_UNKNOWN;
+		cu.pcuMemRelease(mem);
+		return err;
+	}
 
-		result = cuda::CCUDAHandler::cuda.pcuGraphicsSubResourceGetMappedArray(&iit->asImage.array,iit->cudaHandle,*ait,*mit);
-		if (result != CUDA_SUCCESS)
-			return result;
+	if (auto err = reserveAdrressAndMapMemory(&params.ptr, params.granularSize, params.alignment, params.location, mem); CUDA_SUCCESS != err)
+	{
+		CloseHandle(params.osHandle);
+		cu.pcuMemRelease(mem);
+		return err;
 	}
+
+	if (auto err = cu.pcuMemRelease(mem); CUDA_SUCCESS != err)
+	{
+		CloseHandle(params.osHandle);
+		return err;
+	}
+	
+	*outMem = core::smart_refctd_ptr<CCUDASharedMemory>(new CCUDASharedMemory(core::smart_refctd_ptr<CCUDADevice>(this), std::move(params)), core::dont_grab);
+
 	return CUDA_SUCCESS;
 }
-#endif
+
+CUresult CCUDADevice::importGPUSemaphore(core::smart_refctd_ptr<CCUDASharedSemaphore>* outPtr, ISemaphore* sema)
+{
+	if (!sema || !outPtr)
+		return CUDA_ERROR_INVALID_VALUE;
+
+	auto& cu = m_handler->getCUDAFunctionTable();
+	auto handleType = sema->getCreationParams().externalHandleTypes;
+	auto handle = sema->getCreationParams().externalHandle;
+
+	if (!handleType.hasFlags(ISemaphore::EHT_OPAQUE_WIN32) || !handle)
+		return CUDA_ERROR_INVALID_VALUE;
+
+	CUDA_EXTERNAL_SEMAPHORE_HANDLE_DESC desc = {
+		.type = CU_EXTERNAL_SEMAPHORE_HANDLE_TYPE_TIMELINE_SEMAPHORE_WIN32,
+		.handle = {.win32 = {.handle = handle }},
+	};
```

Comment:
ifdef to check for the WIN32 platform, and error "not implemented" for others

Reply:
https://github.com/Devsh-Graphics-Programming/Nabla/pull/517#discussion_r1264589714